File: haohaninfo/即時範例/90.py
# -*- coding: UTF-8 -*-

#取得報價資訊，詳情請查看技巧51
execfile('function.py')

#定義指標變數
MAarray=[]
MAnum=10
lastHMTime=""
lastMAValue=0
lastPrice=0

#假設目前倉位為買方，進場部分請參考第七章
index=1
orderPrice=10300
coverPrice=0

#取得成交資訊
for i in getMatch():		
 MatchInfo=i.split(',')
 HMTime=MatchInfo[0][0:2]+MatchInfo[0][3:5]
 MatchPrice=int(MatchInfo[1])

 if len(MAarray)==0:
  MAarray+=[MatchPrice]
  lastHMTime=HMTime
 else:
  if HMTime==lastHMTime:
   MAarray[-1]=MatchPrice
  elif HMTime!=lastHMTime:
   if len(MAarray)<MAnum:
    MAarray+=[MatchPrice]
   elif len(MAarray)==MAnum:
   	MAarray=MAarray[1:]+[MatchPrice]
   lastHMTime=HMTime
 
 #出場判斷
 if len(MAarray)==MAnum :
  MAValue=float(sum(MAarray))/len(MAarray)
  if lastMAValue==0 and lastPrice==0:
   lastMAValue=MAValue
   lastPrice=MatchPrice
   continue
  print "Price",MatchPrice,"MA",MAValue 
  if index==1:
   if MatchPrice<MAValue and lastPrice>=lastMAValue:
    index=0
    coverPrice=MatchPrice
    print MatchInfo[0],"Order Sell Success!"
    break
  elif index==-1:
   if MatchPrice>MAValue and lastPrice<=lastMAValue:
    index=0
    coverPrice=MatchPrice
    print MatchInfo[0],"Order Buy Success!"
    break
  lastMAValue=MAValue
  lastPrice=MatchPrice 



   

File: haohaninfo/交易實務案例/90.R
#取得即時報價，詳細在技巧51
source("function.R")

#假設目前開倉並且開倉價位在10000
index <- 1

#設定變數初始格式
MAarray <- numeric(0)
MAnum <- 10
lastHMTime <- 0
lastPrice <- 0 
lastMAValue <- 0

while(index !=0 ){

 #取得即時成交資訊
 Mdata<-GetMatchData(DataPath,Date)
 MatchPrice <- as.numeric(Mdata[[1]][2])
 HMTime <- as.numeric(paste0(substr(Mdata[[1]][1],1,2),substr(Mdata[[1]][1],4,5)))

 #更新MA Array
 if(length(MAarray)==0){
  MAarray <- c(MatchPrice,MAarray[1:(MAnum-1)])
  lastHMTime <- HMTime
  next
 }else if (HMTime==lastHMTime){
  MAarray[1] <- MatchPrice
 }else if(HMTime>lastHMTime){    
  MAarray <- c(MatchPrice,MAarray[1:(MAnum-1)])
  lastHMTime <- HMTime
 }

 #當MA滿足10分鐘的資料時，才進行進場判斷
 if(length(na.omit(MAarray))==MAnum){
  #計算MA值
  MAValue <- round(mean(MAarray),2)
  #若沒有前一筆資訊，則無法進行MA穿越判斷
  if (lastMAValue == 0 ){
   lastMAValue <- MAValue
   lastPrice <- MatchPrice
   next
  } 
  #進行判斷，本範例預設進場買單
  if(index==1){
   if(lastPrice >= lastMAValue & MatchPrice < MAValue){
    index <- 0
    print("Cover Buy Success!")
    break
   }
  }else if(index==(-1)){
   if(lastPrice <= lastMAValue & MatchPrice > MAValue){
    index <- 0
    print("Cover Sell Success!")
    break
   }
  }
  #更新最新值
  lastPrice <- MatchPrice
  lastMAValue <- MAValue
 }

}

